File: ralph-workflow/src/guidelines/javascript.rs
//! JavaScript and TypeScript review guidelines
//!
//! Contains guidelines for JavaScript/TypeScript projects including React, Vue, Angular,
//! Node.js backends (Express, Fastify, `NestJS`), and SSR frameworks (Next.js, Nuxt).

use super::base::ReviewGuidelines;
use crate::language_detector::ProjectStack;

/// Add JavaScript-specific guidelines to the review
pub fn add_javascript_guidelines(
    guidelines: ReviewGuidelines,
    stack: &ProjectStack,
) -> ReviewGuidelines {
    let base = ReviewGuidelines {
        quality_checks: guidelines
            .quality_checks
            .into_iter()
            .chain([
                "Use const/let, never var".to_string(),
                "Handle Promise rejections".to_string(),
                "Use async/await over raw Promises".to_string(),
                "Avoid deeply nested callbacks".to_string(),
            ])
            .collect(),
        security_checks: guidelines
            .security_checks
            .into_iter()
            .chain([
                "Sanitize user input before DOM insertion".to_string(),
                "Use Content Security Policy headers".to_string(),
                "Validate data from external APIs".to_string(),
                "Check for prototype pollution vulnerabilities".to_string(),
            ])
            .collect(),
        performance_checks: guidelines
            .performance_checks
            .into_iter()
            .chain([
                "Debounce/throttle frequent event handlers".to_string(),
                "Use appropriate data structures".to_string(),
                "Minimize DOM manipulation".to_string(),
            ])
            .collect(),
        anti_patterns: guidelines
            .anti_patterns
            .into_iter()
            .chain([
                "Avoid == for comparisons (use ===)".to_string(),
                "Don't mutate function arguments".to_string(),
                "Avoid synchronous I/O in Node.js".to_string(),
            ])
            .collect(),
        ..guidelines
    };

    let with_frontend = if stack.frameworks.iter().any(|f| f == "React" || f == "Vue") {
        add_frontend_guidelines(base)
    } else {
        base
    };

    add_framework_guidelines(with_frontend, stack)
}

pub fn add_typescript_guidelines(
    guidelines: ReviewGuidelines,
    stack: &ProjectStack,
) -> ReviewGuidelines {
    let guidelines = add_javascript_guidelines(guidelines, stack);

    ReviewGuidelines {
        quality_checks: guidelines
            .quality_checks
            .into_iter()
            .chain([
                "Use strict TypeScript mode".to_string(),
                "Prefer interfaces over type aliases for objects".to_string(),
                "Use explicit return types for public functions".to_string(),
                "Avoid 'any' type; use 'unknown' if needed".to_string(),
            ])
            .collect(),
        idioms: guidelines
            .idioms
            .into_iter()
            .chain([
                "Use union types for discriminated unions".to_string(),
                "Leverage type inference where clear".to_string(),
                "Use generics appropriately".to_string(),
            ])
            .collect(),
        anti_patterns: guidelines
            .anti_patterns
            .into_iter()
            .chain([
                "Don't use 'as' casts to bypass type checking".to_string(),
                "Avoid non-null assertions (!) without justification".to_string(),
            ])
            .collect(),
        ..guidelines
    }
}

fn add_frontend_guidelines(guidelines: ReviewGuidelines) -> ReviewGuidelines {
    ReviewGuidelines {
        quality_checks: guidelines
            .quality_checks
            .into_iter()
            .chain([
                "Components are properly modularized".to_string(),
                "State management is predictable".to_string(),
                "Accessibility (a11y) is considered".to_string(),
            ])
            .collect(),
        performance_checks: guidelines
            .performance_checks
            .into_iter()
            .chain([
                "Avoid unnecessary re-renders".to_string(),
                "Use lazy loading for large components".to_string(),
                "Optimize bundle size".to_string(),
            ])
            .collect(),
        ..guidelines
    }
}

fn add_framework_guidelines(
    guidelines: ReviewGuidelines,
    stack: &ProjectStack,
) -> ReviewGuidelines {
    stack
        .frameworks
        .iter()
        .fold(guidelines, |acc, framework| match framework.as_str() {
            "React" => add_react_guidelines(acc),
            "Vue" => add_vue_guidelines(acc),
            "Angular" => add_angular_guidelines(acc),
            "Express" | "Fastify" | "NestJS" => add_node_backend_guidelines(acc),
            "Next.js" | "Nuxt" => add_ssr_framework_guidelines(acc),
            _ => acc,
        })
}

fn add_react_guidelines(guidelines: ReviewGuidelines) -> ReviewGuidelines {
    ReviewGuidelines {
        quality_checks: guidelines
            .quality_checks
            .into_iter()
            .chain([
                "Use hooks correctly (rules of hooks)".to_string(),
                "Properly manage component lifecycle".to_string(),
                "Use React.memo for expensive renders".to_string(),
            ])
            .collect(),
        anti_patterns: guidelines
            .anti_patterns
            .into_iter()
            .chain([
                "Avoid prop drilling (use context or state management)".to_string(),
                "Don't mutate state directly".to_string(),
                "Avoid inline functions in render".to_string(),
            ])
            .collect(),
        ..guidelines
    }
}

fn add_vue_guidelines(guidelines: ReviewGuidelines) -> ReviewGuidelines {
    ReviewGuidelines {
        quality_checks: guidelines
            .quality_checks
            .into_iter()
            .chain([
                "Use Composition API for complex logic".to_string(),
                "Follow Vue style guide".to_string(),
                "Use computed properties appropriately".to_string(),
            ])
            .collect(),
        anti_patterns: guidelines
            .anti_patterns
            .into_iter()
            .chain([
                "Avoid watchers when computed works".to_string(),
                "Don't directly mutate props".to_string(),
            ])
            .collect(),
        ..guidelines
    }
}

fn add_angular_guidelines(guidelines: ReviewGuidelines) -> ReviewGuidelines {
    ReviewGuidelines {
        quality_checks: guidelines
            .quality_checks
            .into_iter()
            .chain([
                "Use OnPush change detection where possible".to_string(),
                "Follow Angular style guide".to_string(),
                "Use RxJS operators effectively".to_string(),
            ])
            .collect(),
        security_checks: guidelines
            .security_checks
            .into_iter()
            .chain(["Use Angular's built-in sanitization".to_string()])
            .collect(),
        anti_patterns: guidelines
            .anti_patterns
            .into_iter()
            .chain([
                "Avoid subscribing without unsubscribing".to_string(),
                "Don't use any type".to_string(),
            ])
            .collect(),
        ..guidelines
    }
}

fn add_node_backend_guidelines(guidelines: ReviewGuidelines) -> ReviewGuidelines {
    ReviewGuidelines {
        quality_checks: guidelines
            .quality_checks
            .into_iter()
            .chain([
                "Use middleware pattern effectively".to_string(),
                "Handle errors in middleware".to_string(),
                "Use environment variables for config".to_string(),
            ])
            .collect(),
        security_checks: guidelines
            .security_checks
            .into_iter()
            .chain([
                "Use helmet for security headers".to_string(),
                "Implement rate limiting".to_string(),
                "Validate request body schema".to_string(),
            ])
            .collect(),
        ..guidelines
    }
}

fn add_ssr_framework_guidelines(guidelines: ReviewGuidelines) -> ReviewGuidelines {
    ReviewGuidelines {
        quality_checks: guidelines
            .quality_checks
            .into_iter()
            .chain([
                "Use appropriate rendering strategy (SSR/SSG/ISR)".to_string(),
                "Handle hydration correctly".to_string(),
                "Optimize for Core Web Vitals".to_string(),
            ])
            .collect(),
        performance_checks: guidelines
            .performance_checks
            .into_iter()
            .chain([
                "Minimize client-side JavaScript".to_string(),
                "Use image optimization".to_string(),
            ])
            .collect(),
        ..guidelines
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_javascript_guidelines() {
        let stack = ProjectStack {
            primary_language: "JavaScript".to_string(),
            secondary_languages: vec![],
            frameworks: vec![],
            has_tests: false,
            test_framework: None,
            package_manager: Some("Bun".to_string()),
        };

        let guidelines = add_javascript_guidelines(ReviewGuidelines::default(), &stack);

        assert!(guidelines
            .quality_checks
            .iter()
            .any(|c| c.contains("const/let")));
        assert!(guidelines.anti_patterns.iter().any(|c| c.contains("===")));
    }

    #[test]
    fn test_typescript_react_guidelines() {
        let stack = ProjectStack {
            primary_language: "TypeScript".to_string(),
            secondary_languages: vec!["JavaScript".to_string()],
            frameworks: vec!["React".to_string(), "Next.js".to_string()],
            has_tests: true,
            test_framework: Some("Jest".to_string()),
            package_manager: Some("Bun".to_string()),
        };

        let guidelines = add_typescript_guidelines(ReviewGuidelines::default(), &stack);

        // Should have TypeScript checks
        assert!(guidelines.quality_checks.iter().any(|c| c.contains("any")));
        // Should have React checks
        assert!(guidelines
            .quality_checks
            .iter()
            .any(|c| c.contains("hooks")));
        // Should have Next.js checks
        assert!(guidelines
            .quality_checks
            .iter()
            .any(|c| c.contains("SSR") || c.contains("rendering")));
    }

    #[test]
    fn test_vue_guidelines() {
        let stack = ProjectStack {
            primary_language: "JavaScript".to_string(),
            secondary_languages: vec![],
            frameworks: vec!["Vue".to_string()],
            has_tests: false,
            test_framework: None,
            package_manager: Some("Bun".to_string()),
        };

        let guidelines = add_javascript_guidelines(ReviewGuidelines::default(), &stack);

        assert!(guidelines
            .quality_checks
            .iter()
            .any(|c| c.contains("Composition API")));
    }

    #[test]
    fn test_angular_guidelines() {
        let stack = ProjectStack {
            primary_language: "TypeScript".to_string(),
            secondary_languages: vec![],
            frameworks: vec!["Angular".to_string()],
            has_tests: false,
            test_framework: None,
            package_manager: Some("Bun".to_string()),
        };

        let guidelines = add_typescript_guidelines(ReviewGuidelines::default(), &stack);

        assert!(guidelines
            .quality_checks
            .iter()
            .any(|c| c.contains("OnPush") || c.contains("RxJS")));
    }

    #[test]
    fn test_express_guidelines() {
        let stack = ProjectStack {
            primary_language: "JavaScript".to_string(),
            secondary_languages: vec![],
            frameworks: vec!["Express".to_string()],
            has_tests: false,
            test_framework: None,
            package_manager: Some("Bun".to_string()),
        };

        let guidelines = add_javascript_guidelines(ReviewGuidelines::default(), &stack);

        assert!(guidelines
            .quality_checks
            .iter()
            .any(|c| c.contains("middleware")));
        assert!(guidelines
            .security_checks
            .iter()
            .any(|c| c.contains("helmet")));
    }

    #[test]
    fn test_nextjs_guidelines() {
        let stack = ProjectStack {
            primary_language: "TypeScript".to_string(),
            secondary_languages: vec!["JavaScript".to_string()],
            frameworks: vec!["Next.js".to_string()],
            has_tests: true,
            test_framework: Some("Jest".to_string()),
            package_manager: Some("Bun".to_string()),
        };

        let guidelines = add_typescript_guidelines(ReviewGuidelines::default(), &stack);

        // Should have SSR framework guidelines
        assert!(guidelines
            .quality_checks
            .iter()
            .any(|c| c.contains("SSR") || c.contains("rendering") || c.contains("hydration")));
    }

    #[test]
    fn test_multiple_frameworks_combines_guidelines() {
        let stack = ProjectStack {
            primary_language: "TypeScript".to_string(),
            secondary_languages: vec!["JavaScript".to_string()],
            frameworks: vec!["React".to_string(), "Express".to_string()],
            has_tests: true,
            test_framework: Some("Jest".to_string()),
            package_manager: Some("Bun".to_string()),
        };

        let guidelines = add_typescript_guidelines(ReviewGuidelines::default(), &stack);

        // Should have React-specific checks
        assert!(guidelines
            .quality_checks
            .iter()
            .any(|c| c.contains("hooks")));

        // Should have Express-specific checks
        assert!(guidelines
            .quality_checks
            .iter()
            .any(|c| c.contains("middleware")));
    }
}

File: ralph-python/ralph/guidelines/javascript.py
"""JavaScript-specific review guideline categories.

Ported from the canonical Rust implementation in
``ralph-workflow/src/guidelines/javascript.rs`` and adapted for the Python port.
The module models core JavaScript guidance plus optional framework-specific
extensions for React, Vue, Angular, Node backends, SSR stacks, and TypeScript.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Iterable


class JavaScriptGuidelines:
    """JavaScript and TypeScript review checks.

    Args:
        frameworks: Optional framework names used to add stack-specific guidance.
        typescript: When true, include the TypeScript-specific checks from the
            Rust implementation alongside the JavaScript baseline.
    """

    __slots__ = (
        "anti_patterns",
        "documentation_checks",
        "frameworks",
        "idioms",
        "performance_checks",
        "quality_checks",
        "security_checks",
        "testing_checks",
        "typescript",
    )

    frameworks: tuple[str, ...]
    typescript: bool
    quality_checks: list[str]
    security_checks: list[str]
    performance_checks: list[str]
    testing_checks: list[str]
    documentation_checks: list[str]
    idioms: list[str]
    anti_patterns: list[str]

    def __init__(self, frameworks: Iterable[str] = (), typescript: bool = False) -> None:
        normalized_frameworks = tuple(frameworks)
        self.frameworks = normalized_frameworks
        self.typescript = typescript
        self.quality_checks = [
            "Use const/let, never var.",
            "Handle Promise rejections explicitly.",
            "Use async/await over raw Promises when it improves readability.",
            "Avoid deeply nested callbacks.",
        ]
        self.security_checks = [
            "Sanitize user input before DOM insertion.",
            "Use Content Security Policy headers.",
            "Validate data from external APIs.",
            "Check for prototype pollution vulnerabilities.",
        ]
        self.performance_checks = [
            "Debounce or throttle frequent event handlers.",
            "Use appropriate data structures for lookup and iteration patterns.",
            "Minimize DOM manipulation.",
        ]
        self.testing_checks = [
            "Test async control flow, Promise rejection paths, and error boundaries.",
            "Cover user-driven UI events and state transitions with behavior-focused tests.",
            (
                "Exercise API validation, middleware, and serialization boundaries "
                "in integration tests."
            ),
        ]
        self.documentation_checks = [
            "Document public modules, exported APIs, and non-obvious framework conventions.",
            "Keep README and setup docs aligned with the runtime, build, and framework stack.",
            (
                "Explain security-sensitive data flow, hydration assumptions, "
                "and state ownership rules."
            ),
        ]
        self.idioms = [
            "Prefer immutable updates and small pure helpers where practical.",
            "Use language and framework conventions instead of custom abstractions.",
            "Keep modules focused and exports intentional.",
        ]
        self.anti_patterns = [
            "Avoid == for comparisons (use ===).",
            "Do not mutate function arguments.",
            "Avoid synchronous I/O in Node.js request or worker paths.",
        ]

        if any(framework.casefold() in {"react", "vue"} for framework in normalized_frameworks):
            self._apply_frontend_guidelines()

        for framework in normalized_frameworks:
            self._apply_framework(framework)

        if typescript:
            self._apply_typescript_guidelines()

    def _apply_typescript_guidelines(self) -> None:
        """Add the TypeScript-specific checks from the Rust implementation."""
        self.quality_checks.extend(
            [
                "Use strict TypeScript mode.",
                "Prefer interfaces over type aliases for object shapes.",
                "Use explicit return types for public functions.",
                "Avoid the any type; use unknown when narrowing is required.",
            ]
        )
        self.idioms.extend(
            [
                "Use union types for discriminated unions.",
                "Leverage type inference where intent stays clear.",
                "Use generics appropriately.",
            ]
        )
        self.anti_patterns.extend(
            [
                "Do not use as casts to bypass type checking.",
                "Avoid non-null assertions (!) without justification.",
            ]
        )

    def _apply_frontend_guidelines(self) -> None:
        """Add shared frontend guidance for React and Vue stacks."""
        self.quality_checks.extend(
            [
                "Components are properly modularized.",
                "State management is predictable.",
                "Accessibility (a11y) is considered.",
            ]
        )
        self.performance_checks.extend(
            [
                "Avoid unnecessary re-renders.",
                "Use lazy loading for large components.",
                "Optimize bundle size.",
            ]
        )

    def _apply_framework(self, framework: str) -> None:
        """Add framework-specific guideline extensions."""
        framework_name = framework.casefold()

        if framework_name == "react":
            self.quality_checks.extend(
                [
                    "Use hooks correctly (rules of hooks).",
                    "Properly manage component lifecycle.",
                    "Use React.memo for expensive renders.",
                ]
            )
            self.anti_patterns.extend(
                [
                    "Avoid prop drilling when context or state management fits better.",
                    "Do not mutate state directly.",
                    "Avoid inline functions in render when they create avoidable churn.",
                ]
            )
            return

        if framework_name == "vue":
            self.quality_checks.extend(
                [
                    "Use the Composition API for complex logic.",
                    "Follow the Vue style guide.",
                    "Use computed properties appropriately.",
                ]
            )
            self.anti_patterns.extend(
                [
                    "Avoid watchers when computed properties express the same intent.",
                    "Do not directly mutate props.",
                ]
            )
            return

        if framework_name == "angular":
            self.quality_checks.extend(
                [
                    "Use OnPush change detection where possible.",
                    "Follow the Angular style guide.",
                    "Use RxJS operators effectively.",
                ]
            )
            self.security_checks.append("Use Angular's built-in sanitization.")
            self.anti_patterns.extend(
                [
                    "Avoid subscribing without unsubscribing.",
                    "Do not use the any type.",
                ]
            )
            return

        if framework_name in {"express", "fastify", "nestjs"}:
            self.quality_checks.extend(
                [
                    "Use middleware patterns effectively.",
                    "Handle errors in middleware.",
                    "Use environment variables for configuration.",
                ]
            )
            self.security_checks.extend(
                [
                    "Use helmet or equivalent security headers.",
                    "Implement rate limiting.",
                    "Validate request body schemas.",
                ]
            )
            return

        if framework_name in {"next.js", "nuxt"}:
            self.quality_checks.extend(
                [
                    "Use the appropriate rendering strategy (SSR, SSG, or ISR).",
                    "Handle hydration correctly.",
                    "Optimize for Core Web Vitals.",
                ]
            )
            self.performance_checks.extend(
                [
                    "Minimize client-side JavaScript.",
                    "Use image optimization features.",
                ]
            )

    def summary(self) -> str:
        """Return a short human-readable summary."""
        return (
            f"{len(self.quality_checks)} quality checks, "
            f"{len(self.security_checks)} security checks, "
            f"{len(self.anti_patterns)} anti-patterns"
        )

    def total_checks(self) -> int:
        """Return the total number of configured checks."""
        return sum(
            [
                len(self.quality_checks),
                len(self.security_checks),
                len(self.performance_checks),
                len(self.testing_checks),
                len(self.documentation_checks),
                len(self.idioms),
                len(self.anti_patterns),
            ]
        )


__all__ = ["JavaScriptGuidelines"]
